libfwupd/fwupd-client.c

```diff
@@ -6372,6 +6548,427 @@ fwupd_client_undo_host_security_attr_finish(FwupdClient *self, GAsyncResult *res
 	return g_task_propagate_boolean(G_TASK(res), error);
 }
 
+static void
+fwupd_client_build_report_metadata(JsonBuilder *builder, GHashTable *metadata)
+{
+	GHashTableIter iter;
+	const gchar *key;
+	const gchar *value;
+
+	g_hash_table_iter_init(&iter, metadata);
+	while (g_hash_table_iter_next(&iter, (gpointer *)&key, (gpointer *)&value)) {
+		json_builder_set_member_name(builder, key);
+		json_builder_add_string_value(builder, value);
+	}
+}
+
+/**
+ * fwupd_client_build_report_devices:
+ * @self: a #FwupdClient
+ * @devices: (element-type FwupdDevice): devices
+ * @metadata: (element-type utf8 utf8): attributes
+ * @error: (nullable): optional return location for an error
+ *
+ * Builds a JSON report for the list of devices.
+ *
+ * This function should be called *before* asking the interactive user if they want to upload a
+ * report -- as this function filters devices and may return an error if there is nothing to do.
+ *
+ * Returns: a string, or %NULL if the ID is not present
+ *
+ * Since: 1.9.20
+ **/
+gchar *
+fwupd_client_build_report_devices(FwupdClient *self,
+				  GPtrArray *devices,
+				  GHashTable *metadata,
+				  GError **error)
+{
+	guint cnt = 0;
+	g_autofree gchar *data = NULL;
+	g_autoptr(JsonBuilder) builder = json_builder_new();
+	g_autoptr(JsonGenerator) json_generator = NULL;
+	g_autoptr(JsonNode) json_root = NULL;
+
+	g_return_val_if_fail(FWUPD_IS_CLIENT(self), NULL);
+	g_return_val_if_fail(devices != NULL, NULL);
+	g_return_val_if_fail(metadata != NULL, NULL);
+	g_return_val_if_fail(error == NULL || *error == NULL, NULL);
+
+	json_builder_begin_object(builder);
+	json_builder_set_member_name(builder, "ReportType");
+	json_builder_add_string_value(builder, "device-list");
+	json_builder_set_member_name(builder, "ReportVersion");
+	json_builder_add_int_value(builder, 2);
+	json_builder_set_member_name(builder, "MachineId");
+	json_builder_add_string_value(builder, fwupd_client_get_host_machine_id(self));
+
+	/* this is system metadata not stored in the database */
+	if (g_hash_table_size(metadata) > 0) {
+		json_builder_set_member_name(builder, "Metadata");
+		json_builder_begin_object(builder);
+		fwupd_client_build_report_metadata(builder, metadata);
+		json_builder_end_object(builder);
+	}
+
+	/* devices */
+	json_builder_set_member_name(builder, "Devices");
+	json_builder_begin_array(builder);
+	for (guint i = 0; i < devices->len; i++) {
+		FwupdDevice *dev = g_ptr_array_index(devices, i);
+		if (!fwupd_device_has_flag(dev, FWUPD_DEVICE_FLAG_UPDATABLE) &&
```

Comment:
So I'm wondering if it's worth sending even non-updatable devices. You'll be able to get s very good sampling of extra metadata about the "other" devices in a machine thanks to the CPU plugin.

Reply:
My concern there is then the data collection rationale goes from a clear "we want to engage with vendor to increase LVFS participation" to  something a lot less easy to explain.